File: python/preprocessing.py
"""
preprocessing.py
----------------
Signal preprocessing for the InDEx neutron/gamma acoustic discrimination
pipeline. This is the ONE place the waveform -> feature-vector logic lives
for the Python side; both train_model.py and predict.py import it, so
training and inference are guaranteed to use identical preprocessing.

Steps (matching the report's Section 2.2):
  1. Parse the raw LabVIEW (.lvm) waveform file
  2. Amplitude-normalize the signal
  3. Frame it (1024-sample Hanning window, 80% overlap)
  4. Compute short-time energy (STE) and locate the true nucleation pulse
     via an energy threshold
  5. Compute the STFT, normalize each frame's spectrum, and average over
     the pulse window -> a 512-dimensional feature vector
"""

import numpy as np

# ---- fixed pipeline constants (must match train_model.py / predict.py / the JS mirror) ----
N_FFT = 1024
OVERLAP = 0.8
HOP = int(N_FFT * (1 - OVERLAP))   # 204
THETA = 0.15                        # energy threshold for pulse localization


def read_lvm(path):
    """Parse a LabVIEW Measurement (.lvm) file.

    Returns
    -------
    signal : np.ndarray
        The waveform's amplitude values (the 'Trigger' column).
    delta_t : float
        The sampling interval in seconds (1 / sampling frequency).
    """
    with open(path, "r", errors="ignore") as f:
        lines = f.readlines()

    header_idx = [i for i, l in enumerate(lines) if "End_of_Header" in l]
    if not header_idx:
        raise ValueError(f"{path}: not a recognized .lvm file (no 'End_of_Header' marker)")
    start = header_idx[-1] + 2  # skip the header marker line + the column-name row

    delta_x = None
    for l in lines:
        if l.startswith("Delta_X"):
            delta_x = float(l.split("\t")[1])
            break
    if delta_x is None:
        delta_x = 1e-6  # fall back to 1 MHz if not found

    values = []
    for l in lines[start:]:
        parts = l.strip().split("\t")
        if len(parts) >= 2:
            try:
                values.append(float(parts[1]))
            except ValueError:
                continue

    if len(values) < N_FFT:
        raise ValueError(f"{path}: only {len(values)} samples parsed, need at least {N_FFT}")

    return np.array(values), delta_x


def extract_features(signal, fs=1e6, n_fft=N_FFT, hop=HOP, theta=THETA):
    """Turn a raw waveform into a 512-D short-time spectral-energy feature vector.

    Returns a dict with the feature vector plus everything needed to plot
    the waveform / STE curve / spectrogram (used by predict.py's --plot
    option and by the dashboard).
    """
    xn = signal / np.max(np.abs(signal))                 # amplitude normalization

    n = len(xn)
    starts = list(range(0, n - n_fft, hop))
    frames = np.array([xn[s:s + n_fft] for s in starts])
    win = np.hanning(n_fft)
    xw = frames * win                                      # windowed frames

    E = np.mean(np.abs(xw) ** 2, axis=1)                    # short-time energy
    Ehat = E / np.max(E)                                    # normalized STE

    above = np.where(Ehat >= theta)[0]
    ks, ke = (int(above.min()), int(above.max())) if len(above) else (0, len(starts) - 1)

    X = np.fft.fft(xw, axis=1)[:, : n_fft // 2]             # STFT, keep 0..Nyquist
    P = (np.abs(X) ** 2) / n_fft                             # power spectrum
    Pmax = np.max(P, axis=1, keepdims=True)
    Pmax[Pmax == 0] = 1
    Phat = P / Pmax                                          # per-frame normalization

    feature = np.mean(Phat[ks:ke + 1], axis=0)               # temporal average -> 512-D

    frame_times = np.array(starts) / fs
    return {
        "feature": feature,           # (512,) — what the classifiers consume
        "raw": xn,                    # normalized waveform, for plotting
        "fs": fs,
        "Ehat": Ehat,                 # short-time energy curve
        "frame_times": frame_times,
        "ks": ks, "ke": ke,           # pulse-window frame indices
        "crop_start_t": float(frame_times[ks]),
        "crop_end_t": float(frame_times[ke] + n_fft / fs),
        "stft_P": Phat,                # full spectrogram (frames x 512), for plotting
    }


def featurize_file(path):
    """Convenience wrapper: read a .lvm file straight to a feature vector."""
    signal, dx = read_lvm(path)
    return extract_features(signal, fs=1.0 / dx)

File: js/preprocessing.js
/*
 * preprocessing.js
 * -----------------
 * Client-side mirror of python/preprocessing.py — MUST match it exactly.
 * Takes a raw .lvm file's text content and turns it into the same 512-D
 * feature vector the Python training pipeline produces:
 *   parse -> normalize -> frame (Hanning, 1024, 80% overlap) -> short-time
 *   energy pulse localization (theta=0.15) -> STFT -> per-frame normalize
 *   -> average over the pulse window
 *
 * This has been numerically validated against preprocessing.py: identical
 * inputs produce identical feature vectors to full floating-point
 * precision.
 */

function fftInPlace(re, im){
  const n = re.length;
  for (let i=1, j=0; i<n; i++){
    let bit = n>>1;
    for (; j & bit; bit >>= 1) j ^= bit;
    j ^= bit;
    if (i < j){
      let t=re[i]; re[i]=re[j]; re[j]=t;
      t=im[i]; im[i]=im[j]; im[j]=t;
    }
  }
  for (let len=2; len<=n; len<<=1){
    const ang = -2*Math.PI/len;
    const wr0 = Math.cos(ang), wi0 = Math.sin(ang);
    for (let i=0; i<n; i+=len){
      let curWr=1, curWi=0;
      for (let j=0; j<len/2; j++){
        const ur=re[i+j], ui=im[i+j];
        const vr=re[i+j+len/2]*curWr - im[i+j+len/2]*curWi;
        const vi=re[i+j+len/2]*curWi + im[i+j+len/2]*curWr;
        re[i+j]=ur+vr; im[i+j]=ui+vi;
        re[i+j+len/2]=ur-vr; im[i+j+len/2]=ui-vi;
        const nWr = curWr*wr0 - curWi*wi0;
        const nWi = curWr*wi0 + curWi*wr0;
        curWr=nWr; curWi=nWi;
      }
    }
  }
}

function hanningWindow(N){
  const w = new Float64Array(N);
  for (let n=0;n<N;n++) w[n] = 0.5 - 0.5*Math.cos(2*Math.PI*n/(N-1));
  return w;
}

// ---- LVM text parser (mirrors read_lvm() in the Python pipeline) ----
function parseLVMText(text){
  const lines = text.split(/\r?\n/);
  const headerIdxs = [];
  for (let i=0;i<lines.length;i++) if (lines[i].includes('End_of_Header')) headerIdxs.push(i);
  if (headerIdxs.length === 0){
    throw new Error("Doesn't look like a LabVIEW .lvm file — no 'End_of_Header' marker found.");
  }
  const start = headerIdxs[headerIdxs.length-1] + 2;
  let deltaX = null;
  for (const l of lines){ if (l.startsWith('Delta_X')){ deltaX = parseFloat(l.split('\t')[1]); break; } }
  if (!deltaX || !isFinite(deltaX)) deltaX = 1e-6;
  const vals = [];
  for (let i=start; i<lines.length; i++){
    const parts = lines[i].trim().split('\t');
    if (parts.length >= 2){
      const v = parseFloat(parts[1]);
      if (!isNaN(v)) vals.push(v);
    }
  }
  if (vals.length < 1024){
    throw new Error(`Only found ${vals.length} samples after parsing — need at least 1024 for one STFT frame.`);
  }
  return { vals: Float64Array.from(vals), fs: 1/deltaX };
}

const NFFT = 1024, HOP = 204, THETA = 0.15;
const WIN = hanningWindow(NFFT);

// ---- full pipeline: normalize -> frame -> STE -> localize -> STFT -> 512-D feature ----
function runPipeline(vals, fs){
  let maxAbs = 0;
  for (let i=0;i<vals.length;i++) maxAbs = Math.max(maxAbs, Math.abs(vals[i]));
  if (maxAbs === 0) throw new Error('Waveform is all zeros — nothing to classify.');
  const xn = new Float64Array(vals.length);
  for (let i=0;i<vals.length;i++) xn[i] = vals[i]/maxAbs;

  const starts = [];
  for (let s=0; s<xn.length-NFFT; s+=HOP) starts.push(s);
  const nFrames = starts.length;

  const Ehat = new Float64Array(nFrames);
  const Praw = []; // per-frame normalized power spectra (nFrames x 512)
  let Emax = 0;
  const Eraw = new Float64Array(nFrames);

  for (let k=0; k<nFrames; k++){
    const s = starts[k];
    const re = new Float64Array(NFFT), im = new Float64Array(NFFT);
    let e = 0;
    for (let n=0;n<NFFT;n++){ const v = xn[s+n]*WIN[n]; re[n]=v; e += v*v; }
    e /= NFFT;
    Eraw[k] = e; if (e > Emax) Emax = e;
    fftInPlace(re, im);
    const P = new Float64Array(512);
    let pmax = 0;
    for (let m=0;m<512;m++){ const p = (re[m]*re[m]+im[m]*im[m])/NFFT; P[m]=p; if (p>pmax) pmax=p; }
    if (pmax === 0) pmax = 1;
    for (let m=0;m<512;m++) P[m] /= pmax;
    Praw.push(P);
  }
  for (let k=0;k<nFrames;k++) Ehat[k] = Emax>0 ? Eraw[k]/Emax : 0;

  let ks=-1, ke=-1;
  for (let k=0;k<nFrames;k++) if (Ehat[k] >= THETA){ if (ks===-1) ks=k; ke=k; }
  if (ks === -1){ ks=0; ke=nFrames-1; }

  const feature = new Float64Array(512);
  for (let k=ks; k<=ke; k++) for (let m=0;m<512;m++) feature[m] += Praw[k][m];
  const denom = (ke-ks+1);
  for (let m=0;m<512;m++) feature[m] /= denom;

  const frameTimes = starts.map(s => s/fs);
  const cropStartT = frameTimes[ks];
  const cropEndT = frameTimes[ke] + NFFT/fs;

  return { xn, fs, nFrames, Ehat, frameTimes, ks, ke, cropStartT, cropEndT, feature, Praw };
}

